docs/proposal/aws-multitenancy.md

```diff
@@ -0,0 +1,254 @@
+
+### Goals
+
+1. To enable cluster reconciliation across AWS account boundaries
+2. To maintain backwards compatibility and cause no impact for users who don't intend to make use of this capability
+
+### Non-Goals/Future Work
+
+- Enabling Machines to be provisioned in AWS accounts different than their control planes. This would require adding various AWS infrastructure specification to the AWSMachineTemplate type which currently does not exist.
+
+## Proposal
+
+![Diagram](https://user-images.githubusercontent.com/208667/78036771-b299ad00-7338-11ea-90b2-1a88702fb1dd.png)
+
+The current implementation of CAPA requests a new AWS EC2 and ELB service per cluster and per machine and stores these in fields on the ClusterScope struct. ClusterScopes are reference values which are created per-reconciliation:
+
+```golang
+type ClusterScope struct {  
+   logr.Logger  
+  client      client.Client  
+  patchHelper *patch.Helper  
+  
+  AWSClients  
+  Cluster    *clusterv1.Cluster  
+  AWSCluster *infrav1.AWSCluster  
+}
+```
+
+The field `AWSClients` holds the actual AWS service clients, and is defined like so:
+```golang
+type AWSClients struct {  
+  EC2             ec2iface.EC2API  
+  ELB             elbiface.ELBAPI  
+  SecretsManager  secretsmanageriface.SecretsManagerAPI  
+  ResourceTagging resourcegroupstaggingapiiface.ResourceGroupsTaggingAPIAPI  
+}
+```
+
+The signatures for the functions which create these instances are as follows:
+
+```golang
+func NewClusterScope(params ClusterScopeParams) (*ClusterScope, error) {
+  ...
+  return &ClusterScope{
+    ...
+  }, nil
+}
+```
+
+### Proposed changes:
+#### Data Model
+api/v1alpha3/awscluster_types.go:
+```golang
+type  AWSClusterSpec  struct {
+  + IAMRoleARN *string `json:"iamRoleArn,omitempty"`
+```
+
+#### Code
+- Add a field to the AWSClients struct: `IAMRoleARN *string`
+- Add an optional field to the AWSClusterSpec type named `IAMRoleARN`
+- When creating a ClusterScope:
+	- if the IAMRoleARN has been specified, create EC2 and ELB services with an [STS credentials object configured to use the provided IAM role ARN]([https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/stscreds/](https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/stscreds/))
+	- if the IAMRoleARN has not been specified then create EC2 and ELB services with no credentials specified (preserve existing behavior)
+
+#### Examples
+cluster.yaml
+```yaml
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: Cluster
+metadata:
+  name: my-cluster
+spec:
+  infrastructureRef:
+    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+    kind: AWSCluster
+    name: my-cluster
+```
+awscluster.yaml
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: AWSCluster
+metadata:
+  name: my-cluster
+spec:
+  iamRoleArn: arn:aws:iam::123456789:role/CAPARole
```

Comment:
Only adding IAM role ARN introduces a confused deputy risk, which is important for some of our users. Would need to support external ID as well.